aws/resource_aws_workspaces_directory_test.go

```diff
@@ -516,3 +555,37 @@ resource "aws_workspaces_directory" "main" {
 }
 `, tagKey1, tagValue1, tagKey2, tagValue2)
 }
+
+func testAccWorkspacesDirectoryConfig_ipGroupIds_create(rName string) string {
+	return testAccAwsWorkspacesDirectoryConfig_Prerequisites(rName) + fmt.Sprintf(`
+resource "aws_workspaces_ip_group" "test_alpha" {
+  name        = "%[1]s-alpha"
+}
+
+resource "aws_workspaces_directory" "main" {
+  directory_id = "${aws_directory_service_directory.main.id}"
+  ip_group_ids = ["${aws_workspaces_ip_group.test_alpha.id}"]
+}
+`, rName)
+}
+
+func testAccWorkspacesDirectoryConfig_ipGroupIds_update(rName string) string {
+	return testAccAwsWorkspacesDirectoryConfig_Prerequisites(rName) + fmt.Sprintf(`
+resource "aws_workspaces_ip_group" "test_alpha" {
```

Comment:
@DrFaust92 I caught one issue with workspace directory and IP group mutual dependency. If I remove IP group resource and it's reference in workspaces directory declaration, then I get an error like this:
```
Error: Error Deleting Workspaces IP Group: ResourceAssociatedException: Can not delete IP Group with GroupId - wsipg-5g88k3vby because its attached with these Directories d-9267047639
```
If I understand correctly that happens, because workspace directory depends on IP group implicitly referencing it in `ip_group_ids`, so Terraform deletes dependent resource (IP group) first. But IP group can't be deleted until it is associated with any directory. Sounds like I should declare workspace directory dependency for IP group, but that makes a cycle and violates acyclic graph constraint:
```
resource "aws_workspaces_ip_group" "test_alpha" {
  name        = "%[1]s-alpha"

  depends_on = ["${aws_workspaces_directory.main.id}"]
}

resource "aws_workspaces_directory" "main" {
  directory_id = "${aws_directory_service_directory.main.id}"
  ip_group_ids = ["${aws_workspaces_ip_group.alpha.id}"]
}
```
I have only one idea – disassociate IP group from directory, but it requires to call [DescribeWorkspaceDirectories](https://docs.aws.amazon.com/workspaces/latest/api/API_DescribeWorkspaceDirectories.html) API action and find the directory associated with IP group – [DescribeIpGroups](https://docs.aws.amazon.com/workspaces/latest/api/API_DescribeIpGroups.html) API action doesn't return back refrence to directory. Feels not ideomatic. Could you provide an example for already implenented resources with similar behavior?

Reply:
There is something similar in API Gateway Usage plan resource, care to take a look for inspiration?
im guessing there isn't much to do to beside listing all directories and dissociating all directories from that ip group.

maybe it's worth splitting the association itself to a separate resource. and leave the ip_groups in this resource computed.

wdyt?

Reply:
TBH, I haven't found anything could help in `aws_api_gateway_usage_plan` resource. `UpdateUsagePlan` action handles all change together with `PatchOperations` attribute, which is designed to handle update operations in ordered manner.

---

I doubt `aws_workspaces_ip_group_attachment` makes any difference and solves the cycle issue.
```
resource "aws_workspaces_directory" "main" {
  directory_id = "${aws_directory_service_directory.main.id}"
}

resource "aws_workspaces_ip_group" "test_alpha" {
  name        = "%[1]s-alpha"
}

resource "aws_workspaces_ip_group_attachment" "main" {
  directory_id = "${aws_directory_service_directory.main.id}"
  ip_group_ids = ["${aws_workspaces_ip_group.test_alpha.id}"]
}
```

![WorkspacesIpGroups](https://user-images.githubusercontent.com/892721/89623693-bff21580-d89d-11ea-8fa2-cc06b949635f.png)

As far as attachment refers to directory and ip group resources, they are destroyed first by Terraform. Nor directory, neither ip group can be destroyed until disassociation is completed.

--- 

Seems like we should go with dumb `aws_workspaces_ip_group` resource disassociation on `Delete()`.

@DrFaust92 Thank you for the insight and please don't hesitate to drop any comment regarding my plan 🙇  I'm not a seasoned Terraform provider developer yet, so appreciate any feedback 🙂

Reply:
Take a look a the delete function in usage plan. It deletes the associated stages.

I think your plan seems fine to me, @gdavison want if you want to chime in on this.

Reply:
I've just bumped with another idea – reverse (inject) dependency between resources – drop `aws_workspaces_directory.ip_group_ids` attribute in favor of `aws_workspaces_ip_group.directory_id` one.

```
resource "aws_workspaces_directory" "test" {
  directory_id = "${aws_directory_service_directory.test.id}"
}

resource "aws_workspaces_ip_group" "a" {
  name         = "A"
  directory_id = "${aws_workspaces_directory.test.id}" {
}

resource "aws_workspaces_ip_group" "b" {
  name         = "B"
  directory_id = "${aws_workspaces_directory.test.id}" {
}
```

![WorkspacesIpGroupsReverse](https://user-images.githubusercontent.com/892721/89625926-478d5380-d8a1-11ea-93e7-cf561147e628.png)

I know AWS Terraform provider tries to follow AWS API as close as possible, but this option looks more explicit than implicit group disassociation.

@DrFaust92  @bflad @gdavison What do you think?

Reply:
Seems like a nicer solution but doesn't it mean there needs to be an IP group per directory? Isn't it supposed to be that an IP group can be attached to several directories? (for less duplication)

Reply:
Good point, indeed IP group is supposed to be associated with many directories. Let's allow multiple values within `directory_ids` attribute:
```
resource "aws_workspaces_directory" "x" {
  directory_id = "${aws_directory_service_directory.x.id}"
}

resource "aws_workspaces_directory" "y" {
  directory_id = "${aws_directory_service_directory.y.id}"
}

resource "aws_workspaces_ip_group" "a" {
  directory_ids = [
    "${aws_workspaces_directory.x.id}",
  ]
}

resource "aws_workspaces_ip_group" "b" {
  directory_ids = [
    "${aws_workspaces_directory.x.id}",
    "${aws_workspaces_directory.y.id}",
  ]
}
```
Doesn't look cool, but solves the problem. @DrFaust92 Is it better from your perspective?

Reply:
It looks better, I'm just wondering if it becomes too unintuitive for users. Let's wait for maintainers to comment as well.

Reply:
@gdavison Please could you suggest the best option here?

Reply:
Thanks for working on this @Tensho. I personally prefer the attachment resource option, this would be more intuitive and most similar to other resources

Reply:
@bensunny1 TBH, I can't grasp how attachment may solve the [problem I mentioned above in this thread](https://github.com/terraform-providers/terraform-provider-aws/pull/14451/files#r466892431). If I remove IP group attachment and IP group from the configuration **all together**, then both resources will be marked as orphaned and deleted in undesired order (first IP group, then IP group attachment) by Terraform. I guess it's OK for other attachments and target resources, e.g. `aws_iam_role_policy_attachment` and `aws_iam_role`/`aws_iam_policy`, but it doesn't work for IP group API. Or maybe I don't understand how Terraform builds the graph nodes for `destroy` operations...

@apparentlymart @bflad Please help us to sort out this chicken-egg question 🙇

Reply:
@Tensho You're right, https://github.com/terraform-providers/terraform-provider-aws/pull/14451#discussion_r466961078 i think is a pretty good idea and deals with the problem

Reply:
I think the best option here is to add the `DisassociateIpGroups()` call to the `aws_workspaces_ip_group` delete operation as well. There isn't currently a way to model these mutual relationships in Terraform, so we're stuck with brute-forcing it.

Reply:
@gdavison Got it. Added ip group disassociation from a directory on delete.